File: 总题库/394.DecodingString.py
class Solution:
    def decodeString(self, s: str) -> str:
        times = []
        chars = []
        number = 0
        res = ''

        for i in range(len(s)):
            if '0' <= s[i] and s[i] <= '9':
                number = number * 10 + int(s[i])
            elif s[i] == ']':
                temp = ''
                final = ''
                index = len(chars) - 1
                while chars[-1] != '[':
                    temp = chars[-1] + temp
                    index -= 1
                    chars.pop()
                chars.pop()
                curTimes = times[-1]
                times.pop()
                while curTimes != 0:
                    final += temp
                    curTimes -= 1
                chars.append(final)
            else:
                if number:
                    times.append(number)
                    number = 0
                chars.append(s[i])
        for i in range(len(chars)):
            res += chars[i];

        return res

File: 总题库/394.DecodingString.cpp
class Solution {
public:
    string decodeString(string s) {
        stack<int> times;
        vector<string> chars;
        int number = 0;
        string res = "";
        for (int i = 0; i < s.length(); i++) {
            if (isdigit(s[i])) {
                number = number * 10 + (s[i] - '0');
            }
            else if (s[i] == ']') {
                string temp = "", final = "";
                int index = chars.size() - 1;
                while (chars.back() != "[") {
                    temp = chars.back() + temp;
                    index--;
                    chars.pop_back();
                }
                chars.pop_back();
                // reverse(temp.begin(), temp.end());
                int curTimes = times.top();
                times.pop();
                while (curTimes--) {
                    final += temp;
                }
                chars.emplace_back(final);
            }
            else {
                if (number) {
                    times.push(number);
                    number = 0;
                }
                string temp = "";
                chars.push_back(temp + s[i]);
            }
        }
        for (int i = 0; i < chars.size(); i++) {
            res += chars[i];
        }
        return res;
    }
};